lib/autoyast.pm

```diff
@@ -207,125 +207,287 @@ sub init_autoyast_profile {
 
 =cut
 sub validate_autoyast_profile {
     my $profile = shift;
 
-    my $xpc         = get_xpc(init_autoyast_profile());    # get XPathContext
-    my $expressions = [];
+    my $xpc         = get_xpc(init_autoyast_profile());                              # get XPathContext
     my $errors      = [];
-    generate_expressions(node => $profile, exp => "/ns:profile",
-        expressions => $expressions, errors => $errors);
+    my $expressions = [map { '/ns:profile' . $_ } generate_expressions($profile)];
     run_expressions(xpc => $xpc, expressions => $expressions, errors => $errors);
     my $report = create_report(errors => $errors, expressions => $expressions);
     record_info('Summary', $report);
-    die "Found errors on validation of AutoYaST profile, please check Summary report" if @{$errors};
+    die 'Found errors on validation of AutoYaST profile, ' .
+      'please check Summary report' if (@{$errors});
 }
 
-sub generate_expressions {
-    my (%args) = @_;
+=head2 is_processable
 
-    my $node        = $args{node};
-    my $exp         = $args{exp};
-    my $expressions = $args{expressions};
-    my $errors      = $args{errors};
+    is_processable($node);
+
+  A node is considered 'processable' when:
+   - is a simple key-value pair where value is text:
+     quotas: 'true'
+   - has properties:
+     quotas:
+       _t: boolean
+       __text: 'true'
+  In other words, a node is not processable when just needs to be traversed in the YAML.
+
+=cut
+sub is_processable {
+    my $node = shift;
+    return (!ref $node ||
+          (ref $node eq 'HASH' && has_properties($node)));
+}
+
+=head2 has_properties
+
+  has_properties($node);
+
+- A XML node can be specified in YAML along with its attributes from XML:
+    XML:  <subvolumes t="list">
+    YAML: subvolumes:
+            _t: list
+    XPATH: ns:subvolumes[@t='list']
+
+- XPATH functions can be used:
+
+    'count': when you want to explicitly count the element in a list
+    XML:  <subvolumes t="list">
+            <subvolume t="map">
+    YAML: subvolumes:
+            __count: 8
+            subvolume:
+                ...
+    XPATH: ns:subvolumes[count(ns:subvolume)=8]
+
+    'text': when you want to check some attribute and the text itself
+    XML:  <quotas t="boolean">true</quotas>
+    YAML: quotas:
+            _t: boolean
+            __text: true
+    XPATH: ns:quotas[text()='true' and @t='boolean']
+
+    'descendant': by default the algorithm will try to identify an element
+    in a list using direct children, but with this property is it possible to
+    add children which are descendant but not necessarily direct children.
+    It is useful when XML structure are almost exactly the same for different
+    list items and the only difference is found in more deeper descendants,
+    therefore avoiding to return multiple result for an XPATH expression.
+
+    YAML:
+            drive:
+            - label: 
+                _descendant: any
```

Comment:
I wonder if we want to make it default, so do not have to define it in each yaml file.

Reply:
for now it only is needed in a particular scenario, multi_btrfs, in general default with 'direct children' works and when one checks the yaml, mimic better what is in the xml. If we would make the default 'any descendant' I think it would be more difficult to read the yaml, because someone could use always internal properties of a list item which break the idea of mirrow xml/yaml and also we could shadow something internal with the same if it is a direct node, but just thinking loud, because of that I chose the default 'direct child' is more simple to understand I believe.

Reply:
Ah, I've mixed it up with something else. It should not be default and tbh, I would not even have this feature, due to the reasons you have stated. As for multibtrfs case, would it work if we add `drive` node without any details, and look for any direct children node? Let's discuss it quickly after the daily.

Reply:
the initial implementation was based in multi_btrfs, for that reason I ended up creating those magic keys, this time I worked in reverse order (which is better because it is thought more for the general cases not the exception), but I faced the problem as well. We have there [this](https://openqa.suse.de/tests/5585546/file/clone-autoinst.xml)
```xml
<drive t="map">
    <btrfs_options t="map"></btrfs_options>
    <device>btrfs_57</device>
    <disklabel>none</disklabel>
    <enable_snapshots t="boolean">false</enable_snapshots>
    <partitions t="list"></partitions>
    <type t="symbol">CT_BTRFS</type>
    <use>all</use>
</drive>
<drive t="map">
    <btrfs_options t="map">
        <data_raid_level>raid0</data_raid_level>
        <metadata_raid_level>raid1</metadata_raid_level>
    </btrfs_options>
    <device>btrfs_75</device>
    <disklabel>none</disklabel>
    <partitions t="list"></partitions>
    <type t="symbol">CT_BTRFS</type>
    <use>all</use>
</drive>
```
There those 'btrfs_NN' depends on the run, they are not fixed so the rest of information is the same for those last 'drive' nodes (previous for 'drive' nodes don't have any problem). The only way to know which one is it and not return two nodes as result of the expression is to go inside 'partitions' node of each one to check for more values.
Let's discuss in the daily, just want to share exact example.